agent/tcs/client/client.go

```diff
@@ -0,0 +1,138 @@
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4"
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/aws/amazon-ecs-agent/agent/stats"
+	"github.com/aws/amazon-ecs-agent/agent/tcs/model/ecstcs"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient"
+	"github.com/gorilla/websocket"
+)
+
+var log = logger.ForModule("tcs client")
+
+// clientServer implements wsclient.ClientServer interface for metrics backend.
+type clientServer struct {
+	statsEngine  stats.Engine
+	publishTimer *timer
+	wsclient.ClientServerImpl
+	signer authv4.HttpSigner
+}
+
+// New returns a client/server to bidirectionally communicate with the backend.
+// The returned struct should have both 'Connect' and 'Serve' called upon it
+// before being used.
+func New(url string, region string, credentialProvider credentials.AWSCredentialProvider, acceptInvalidCert bool, statsEngine stats.Engine, publishMetricsInterval time.Duration) wsclient.ClientServer {
+	cs := &clientServer{
+		statsEngine:  statsEngine,
+		publishTimer: newTimer(publishMetricsInterval, publishMetrics),
+		signer:       authv4.NewHttpSigner(region, wsclient.ServiceName, credentialProvider, nil),
+	}
+	cs.URL = url
+	cs.Region = region
+	cs.CredentialProvider = credentialProvider
+	cs.AcceptInvalidCert = acceptInvalidCert
+	cs.ServiceError = &tcsError{}
+	cs.RequestHandlers = make(map[string]wsclient.RequestHandler)
+	cs.TypeDecoder = &TcsDecoder{}
+	return cs
+}
+
+// Serve begins serving requests using previously registered handlers (see
+// AddRequestHandler). All request handlers should be added prior to making this
+// call as unhandled requests will be discarded.
+func (cs *clientServer) Serve() error {
+	log.Debug("Starting websocket poll loop")
+	if cs.Conn == nil {
+		return errors.New("nil connection")
+	}
+
+	// Start the timer function to publish metrics to the backend.
+	if cs.statsEngine != nil {
+		cs.publishTimer.start(cs)
+	}
+
+	return cs.ConsumeMessages()
+}
+
+// MakeRequest makes a request using the given input. Note, the input *MUST* be
+// a pointer to a valid backend type that this client recognises
+func (cs *clientServer) MakeRequest(input interface{}) error {
+	payload, err := cs.CreateRequestMessage(input)
+	if err != nil {
+		return err
+	}
+
+	data := cs.signRequest(payload)
+
+	// Over the wire we send something like
+	// {"type":"AckRequest","message":{"messageId":"xyz"}}
+	// return cs.Conn.WriteMessage(websocket.TextMessage, send)
+	return cs.Conn.WriteMessage(websocket.TextMessage, data)
+}
+
+func (cs *clientServer) signRequest(payload []byte) []byte {
+	signer := authv4.NewHttpSigner(cs.Region, "ecs", cs.CredentialProvider, nil)
+	reqBody := bytes.NewBuffer(payload)
+	// NewRequest never returns an error if the url parses and we just verified
+	// it did above
+	request, _ := http.NewRequest("GET", cs.URL, reqBody)
+	signer.SignHttpRequest(request)
+
+	var data []byte
```

Comment:
I think you can simplify this code a bit by using request.Header.Write. [Here's](http://play.golang.org/p/ELaChzwh75) a play showing what I'm thinking of.